File: 2023/day2/main.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func part1(input string) int {
	lines := strings.Split(input, "\n")
	lines = lines[:len(lines)-1]

	cubes := map[string]int{
		"red":   12,
		"green": 13,
		"blue":  14,
	}

	sum := 0
	for i, line := range lines {
		s := strings.SplitN(line, ": ", 2)
		sets := strings.Split(s[1], "; ")

		possible := true
		for _, set := range sets {
			showed := strings.Split(set, ", ")
			for _, c := range showed {
				cc := strings.SplitN(c, " ", 2)
				count, _ := strconv.Atoi(cc[0])
				color := cc[1]
				if cubes[color] < count {
					possible = false
					break
				}
			}
		}
		if possible {
			sum += i + 1
		}
	}
	return sum
}

func part2(input string) int {
	lines := strings.Split(input, "\n")
	lines = lines[:len(lines)-1]

	total := 0
	for _, line := range lines {
		s := strings.SplitN(line, ": ", 2)
		sets := strings.Split(s[1], "; ")

		red, green, blue := 0, 0, 0
		for _, set := range sets {
			cubes := strings.Split(set, ", ")
			for _, c := range cubes {
				cc := strings.SplitN(c, " ", 2)
				n, _ := strconv.Atoi(cc[0])
				color := cc[1]

				switch color {
				case "red":
					if n > red {
						red = n
					}
				case "green":
					if n > green {
						green = n
					}
				case "blue":
					if n > blue {
						blue = n
					}
				}
			}
		}

		total += red * green * blue
	}

	return total
}

func main() {
	bytes, err := os.ReadFile("inputs/2_input.txt")
	input := string(bytes)

	if err != nil {
		log.Fatal(err)
		return
	}

	res := part1(input)
	fmt.Println(res)
	fmt.Println(part2(input))
}

File: 2023/day2/main.rs
use std::{collections::HashMap, fs, io};

fn main() -> io::Result<()> {
    let input = fs::read_to_string("inputs/2_input.txt")?;

    println!("part 1: {}", part_1(&input));
    println!("part 2: {}", part_2(&input));
    Ok(())
}

fn part_1(input: &str) -> usize {
    let cubes: HashMap<&str, usize> = HashMap::from([("red", 12), ("green", 13), ("blue", 14)]);

    input
        .lines()
        .filter_map(|line| {
            let (id, sets) = line.split_once(": ").unwrap();
            let mut sets = sets.split("; ");

            let impossible_set = sets.find(|set| {
                let set = set.split(", ");
                let mut set = set.filter(|cube| {
                    if let Some((n, color)) = cube.split_once(" ") {
                        if n.parse::<usize>().unwrap() > *cubes.get(color).unwrap() {
                            return true;
                        }
                    }
                    false
                });
                set.next().is_some()
            });

            if None == impossible_set {
                return Some(
                    id.split_whitespace()
                        .last()
                        .unwrap()
                        .parse::<usize>()
                        .unwrap(),
                );
            }
            None
        })
        .sum()
}

fn part_2(input: &str) -> usize {
    input
        .lines()
        .map(|line| {
            let (_, sets) = line.split_once(": ").unwrap();
            let sets = sets.split("; ");
            let (mut red, mut green, mut blue) = (0, 0, 0);
            for set in sets {
                for cube in set.split(", ") {
                    if let Some((n, color)) = cube.split_once(" ") {
                        let n = n.parse().unwrap();
                        match color {
                            "red" if n > red => red = n,
                            "green" if n > green => green = n,
                            "blue" if n > blue => blue = n,
                            _ => (),
                        };
                    }
                }
            }
            red * green * blue
        })
        .sum()
}
